Basic auto formula for forward grad

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43718 make jvp use forward mode by default
* #43717 Add gradcheck for forward grad
* #43716 Add many forward grad
* **#43715 Basic auto formula for forward grad**
* #43714 make sure SavedVariable propagate fw_grad properly
* #43713 Add forward grad mode in python and c++
* #43712 Add fw_grad field and accessors

As an example, below are the generated code for the two functions changed here:

For abs, it adds the following block of code at the end of the function (just before the return statement):
```cpp
  if (FwGradMode::is_enabled() && (isFwGradDefined(self))) {
      NoFwGradGuard guard;
      auto self_fw_grad = toLegacyFwGrad(self);
      auto result_new_fw_grad = self_fw_grad * self.sign();
      result.set_fw_grad(result_new_fw_grad);
  }
```
Note that the inplace version of abs, `abs_`  that uses the same entry in derivatives.yaml gets the following line added:
```cpp
  TORCH_CHECK(!(FwGradMode::is_enabled() && (isFwGradDefined(self))), "Trying to use forward prop with abs_ (because it is inplace) that does not support it.");
```

For as_strided, it adds the following:
```cpp
  if (FwGradMode::is_enabled() && (isFwGradDefined(self))) {
      NoFwGradGuard guard;
      auto self_fw_grad = toLegacyFwGrad(self);
      auto result_new_fw_grad = at::as_strided(self_fw_grad, size, stride, storage_offset);
      result.set_fw_grad(result_new_fw_grad);
  }
```

The full code for the abs function is thus:
```cpp
Tensor abs(const Tensor & self) {
  auto& self_ = unpack(self, "self", 0);
  std::shared_ptr<AbsBackward> grad_fn;
  if (compute_requires_grad( self )) {
    grad_fn = std::shared_ptr<AbsBackward>(new AbsBackward(), deleteNode);
    grad_fn->set_next_edges(collect_next_edges( self ));
    grad_fn->self_ = SavedVariable(self, false);
  }
  #ifndef NDEBUG
  c10::optional<Storage> self__storage_saved =
    self_.has_storage() ? c10::optional<Storage>(self_.storage()) : c10::nullopt;
  c10::intrusive_ptr<TensorImpl> self__impl_saved;
  if (self_.defined()) self__impl_saved = self_.getIntrusivePtr();
  #endif
  auto tmp = ([&]() {
    at::AutoNonVariableTypeMode non_var_type_mode(true);
    return at::abs(self_);
  })();
  auto result = std::move(tmp);
  #ifndef NDEBUG
  if (self__storage_saved.has_value())
    AT_ASSERT(self__storage_saved.value().is_alias_of(self_.storage()));
  if (self__impl_saved) AT_ASSERT(self__impl_saved == self_.getIntrusivePtr());
  #endif
  if (grad_fn) {
      set_history(flatten_tensor_args( result ), grad_fn);
  }
  if (FwGradMode::is_enabled() && (isFwGradDefined(self))) {
      NoFwGradGuard guard;
      auto self_fw_grad = toLegacyFwGrad(self);
      auto result_new_fw_grad = self_fw_grad * self.sign();
      result.set_fw_grad(result_new_fw_grad);
  }
  return result;
}
```

Differential Revision: [D23454143](https://our.internmc.facebook.com/intern/diff/D23454143)

albanD committed:
Basic auto formula for forward grad

[ghstack-poisoned]

albanD committed:
Update on "Basic auto formula for forward grad"




[ghstack-poisoned]

zou3519 commented:
Could you copy & paste some generated code for the forward AD so that it's easier to imagine what the codegen produdes?

zou3519 commented on tools/autograd/derivatives.yaml:
```diff
@@ -249,6 +250,7 @@
 
 - name: as_strided(Tensor(a) self, int[] size, int[] stride, int? storage_offset=None) -> Tensor(a)
   self: as_strided_backward(grad, TensorGeometry(self), size, stride, storage_offset)
+  result: linear
```
Add a test for `element_wise` / `linear`?

albanD replied:
You mean to make sure the gradients are correct?
I prefer to delay that until there is gradcheck higher in the stack (I'll merge the whole thing as at once. So this won't be in master without being tested!)

albanD committed:
Update on "Basic auto formula for forward grad"




[ghstack-poisoned]

albanD committed:
Update on "Basic auto formula for forward grad"




[ghstack-poisoned]

albanD committed:
Update on "Basic auto formula for forward grad"




[ghstack-poisoned]

albanD committed:
Update on "Basic auto formula for forward grad"




[ghstack-poisoned]

ezyang commented:
I'm not a big fan of special keywords that look as if they could be regular expressions; there is no syntactic indication that the keyword is special; it could simply just be a plain old variable reference (but it's not). Maybe look and see if there is some extra syntactic marker, something that isn't valid C++ code otherwise, which you can use to indicate something special is going on here.

albanD commented:
I agree that this is not the best. I followed the practice of the `non_differentiable` special word we already have in this file.

> extra syntactic marker, something that isn't valid C++ code otherwise, 

Something that would look like a template maybe?
```
  result: ${auto_element_wise}
```

ezyang commented:
> Something that would look like a template maybe?

Is it actually implemented as template transclusion, though? ;)

albanD commented:
Not it's implemented like the existing "non_differentiable" by doing string comparison...

albanD committed:
Update on "Basic auto formula for forward grad"




[ghstack-poisoned]

albanD committed:
Update on "Basic auto formula for forward grad"




[ghstack-poisoned]

albanD committed:
Update on "Basic auto formula for forward grad"




[ghstack-poisoned]

albanD committed:
Update on "Basic auto formula for forward grad"




[ghstack-poisoned]

albanD committed:
Update on "Basic auto formula for forward grad"


As an example, below are the generated code for the two functions changed here:

For abs, it adds the following block of code at the end of the function (just before the return statement):
```cpp
  if (FwGradMode::is_enabled() && (isFwGradDefined(self))) {
      NoFwGradGuard guard;
      auto self_fw_grad = toLegacyFwGrad(self);
      auto result_new_fw_grad = self_fw_grad * self.sign();
      result.set_fw_grad(result_new_fw_grad);
  }
```
Note that the inplace version of abs, `abs_`  that uses the same entry in derivatives.yaml gets the following line added:
```cpp
  TORCH_CHECK(!(FwGradMode::is_enabled() && (isFwGradDefined(self))), "Trying to use forward prop with abs_ (because it is inplace) that does not support it.");
```

For as_strided, it adds the following:
```cpp
  if (FwGradMode::is_enabled() && (isFwGradDefined(self))) {
      NoFwGradGuard guard;
      auto self_fw_grad = toLegacyFwGrad(self);
      auto result_new_fw_grad = at::as_strided(self_fw_grad, size, stride, storage_offset);
      result.set_fw_grad(result_new_fw_grad);
  }
```

The full code for the abs function is thus:
```cpp
Tensor abs(const Tensor & self) {
  auto& self_ = unpack(self, "self", 0);
  std::shared_ptr<AbsBackward> grad_fn;
  if (compute_requires_grad( self )) {
    grad_fn = std::shared_ptr<AbsBackward>(new AbsBackward(), deleteNode);
    grad_fn->set_next_edges(collect_next_edges( self ));
    grad_fn->self_ = SavedVariable(self, false);
  }
  #ifndef NDEBUG
  c10::optional<Storage> self__storage_saved =
    self_.has_storage() ? c10::optional<Storage>(self_.storage()) : c10::nullopt;
  c10::intrusive_ptr<TensorImpl> self__impl_saved;
  if (self_.defined()) self__impl_saved = self_.getIntrusivePtr();
  #endif
  auto tmp = ([&]() {
    at::AutoNonVariableTypeMode non_var_type_mode(true);
    return at::abs(self_);
  })();
  auto result = std::move(tmp);
  #ifndef NDEBUG
  if (self__storage_saved.has_value())
    AT_ASSERT(self__storage_saved.value().is_alias_of(self_.storage()));
  if (self__impl_saved) AT_ASSERT(self__impl_saved == self_.getIntrusivePtr());
  #endif
  if (grad_fn) {
      set_history(flatten_tensor_args( result ), grad_fn);
  }
  if (FwGradMode::is_enabled() && (isFwGradDefined(self))) {
      NoFwGradGuard guard;
      auto self_fw_grad = toLegacyFwGrad(self);
      auto result_new_fw_grad = self_fw_grad * self.sign();
      result.set_fw_grad(result_new_fw_grad);
  }
  return result;
}
```

[ghstack-poisoned]

albanD committed:
Update on "Basic auto formula for forward grad"


As an example, below are the generated code for the two functions changed here:

For abs, it adds the following block of code at the end of the function (just before the return statement):
```cpp
  if (FwGradMode::is_enabled() && (isFwGradDefined(self))) {
      NoFwGradGuard guard;
      auto self_fw_grad = toLegacyFwGrad(self);
      auto result_new_fw_grad = self_fw_grad * self.sign();
      result.set_fw_grad(result_new_fw_grad);
  }
```
Note that the inplace version of abs, `abs_`  that uses the same entry in derivatives.yaml gets the following line added:
```cpp
  TORCH_CHECK(!(FwGradMode::is_enabled() && (isFwGradDefined(self))), "Trying to use forward prop with abs_ (because it is inplace) that does not support it.");
```

For as_strided, it adds the following:
```cpp
  if (FwGradMode::is_enabled() && (isFwGradDefined(self))) {
      NoFwGradGuard guard;
      auto self_fw_grad = toLegacyFwGrad(self);
      auto result_new_fw_grad = at::as_strided(self_fw_grad, size, stride, storage_offset);
      result.set_fw_grad(result_new_fw_grad);
  }
```

The full code for the abs function is thus:
```cpp
Tensor abs(const Tensor & self) {
  auto& self_ = unpack(self, "self", 0);
  std::shared_ptr<AbsBackward> grad_fn;
  if (compute_requires_grad( self )) {
    grad_fn = std::shared_ptr<AbsBackward>(new AbsBackward(), deleteNode);
    grad_fn->set_next_edges(collect_next_edges( self ));
    grad_fn->self_ = SavedVariable(self, false);
  }
  #ifndef NDEBUG
  c10::optional<Storage> self__storage_saved =
    self_.has_storage() ? c10::optional<Storage>(self_.storage()) : c10::nullopt;
  c10::intrusive_ptr<TensorImpl> self__impl_saved;
  if (self_.defined()) self__impl_saved = self_.getIntrusivePtr();
  #endif
  auto tmp = ([&]() {
    at::AutoNonVariableTypeMode non_var_type_mode(true);
    return at::abs(self_);
  })();
  auto result = std::move(tmp);
  #ifndef NDEBUG
  if (self__storage_saved.has_value())
    AT_ASSERT(self__storage_saved.value().is_alias_of(self_.storage()));
  if (self__impl_saved) AT_ASSERT(self__impl_saved == self_.getIntrusivePtr());
  #endif
  if (grad_fn) {
      set_history(flatten_tensor_args( result ), grad_fn);
  }
  if (FwGradMode::is_enabled() && (isFwGradDefined(self))) {
      NoFwGradGuard guard;
      auto self_fw_grad = toLegacyFwGrad(self);
      auto result_new_fw_grad = self_fw_grad * self.sign();
      result.set_fw_grad(result_new_fw_grad);
  }
  return result;
}
```

[ghstack-poisoned]

albanD committed:
Update on "Basic auto formula for forward grad"


As an example, below are the generated code for the two functions changed here:

For abs, it adds the following block of code at the end of the function (just before the return statement):
```cpp
  if (FwGradMode::is_enabled() && (isFwGradDefined(self))) {
      NoFwGradGuard guard;
      auto self_fw_grad = toLegacyFwGrad(self);
      auto result_new_fw_grad = self_fw_grad * self.sign();
      result.set_fw_grad(result_new_fw_grad);
  }
```
Note that the inplace version of abs, `abs_`  that uses the same entry in derivatives.yaml gets the following line added:
```cpp
  TORCH_CHECK(!(FwGradMode::is_enabled() && (isFwGradDefined(self))), "Trying to use forward prop with abs_ (because it is inplace) that does not support it.");
```

For as_strided, it adds the following:
```cpp
  if (FwGradMode::is_enabled() && (isFwGradDefined(self))) {
      NoFwGradGuard guard;
      auto self_fw_grad = toLegacyFwGrad(self);
      auto result_new_fw_grad = at::as_strided(self_fw_grad, size, stride, storage_offset);
      result.set_fw_grad(result_new_fw_grad);
  }
```

The full code for the abs function is thus:
```cpp
Tensor abs(const Tensor & self) {
  auto& self_ = unpack(self, "self", 0);
  std::shared_ptr<AbsBackward> grad_fn;
  if (compute_requires_grad( self )) {
    grad_fn = std::shared_ptr<AbsBackward>(new AbsBackward(), deleteNode);
    grad_fn->set_next_edges(collect_next_edges( self ));
    grad_fn->self_ = SavedVariable(self, false);
  }
  #ifndef NDEBUG
  c10::optional<Storage> self__storage_saved =
    self_.has_storage() ? c10::optional<Storage>(self_.storage()) : c10::nullopt;
  c10::intrusive_ptr<TensorImpl> self__impl_saved;
  if (self_.defined()) self__impl_saved = self_.getIntrusivePtr();
  #endif
  auto tmp = ([&]() {
    at::AutoNonVariableTypeMode non_var_type_mode(true);
    return at::abs(self_);
  })();
  auto result = std::move(tmp);
  #ifndef NDEBUG
  if (self__storage_saved.has_value())
    AT_ASSERT(self__storage_saved.value().is_alias_of(self_.storage()));
  if (self__impl_saved) AT_ASSERT(self__impl_saved == self_.getIntrusivePtr());
  #endif
  if (grad_fn) {
      set_history(flatten_tensor_args( result ), grad_fn);
  }
  if (FwGradMode::is_enabled() && (isFwGradDefined(self))) {
      NoFwGradGuard guard;
      auto self_fw_grad = toLegacyFwGrad(self);
      auto result_new_fw_grad = self_fw_grad * self.sign();
      result.set_fw_grad(result_new_fw_grad);
  }
  return result;
}
```

Differential Revision: [D23454143](https://our.internmc.facebook.com/intern/diff/D23454143)

[ghstack-poisoned]
